exchanges/ftx/ftx_websocket.go

```diff
@@ -0,0 +1,485 @@
+func (f *FTX) Unsubscribe(channelToSubscribe wshandler.WebsocketChannelSubscription) error {
+	var unSub WsSub
+	a, err := f.GetPairAssetType(channelToSubscribe.Currency)
+	if err != nil {
+		return err
+	}
+	unSub.Operation = unsubscribe
+	unSub.Channel = channelToSubscribe.Channel
+	unSub.Market = f.FormatExchangeCurrency(channelToSubscribe.Currency, a).String()
+	return f.WebsocketConn.SendJSONMessage(unSub)
+}
+
+// WsProcessUpdateOB processes an update on the orderbook
+func (f *FTX) WsProcessUpdateOB(data *WsOrderbookData, p currency.Pair, a asset.Item) error {
+	update := wsorderbook.WebsocketOrderbookUpdate{
+		Asset:      a,
+		Pair:       p,
+		UpdateTime: timestampFromFloat64(data.Time),
+	}
+
+	var err error
+	for x := range data.Bids {
+		update.Bids = append(update.Bids, orderbook.Item{
+			Price:  data.Bids[x][0],
+			Amount: data.Bids[x][1],
+		})
+	}
+	for x := range data.Asks {
+		update.Asks = append(update.Asks, orderbook.Item{
+			Price:  data.Asks[x][0],
+			Amount: data.Asks[x][1],
+		})
+	}
+
+	err = f.Websocket.Orderbook.Update(&update)
+	if err != nil {
+		return err
+	}
+
+	updatedOb := f.Websocket.Orderbook.GetOrderbook(p, a)
+	checksum := f.CalcUpdateOBChecksum(updatedOb)
+
+	if checksum != data.Checksum {
+		log.Warnf(log.ExchangeSys, "%s checksum failure for item %s",
+			f.Name,
+			p)
+		return errors.New("checksum failed")
+	}
+	f.Websocket.DataHandler <- wshandler.WebsocketOrderbookUpdate{
+		Exchange: f.Name,
+		Asset:    a,
+		Pair:     p,
+	}
+
+	return nil
+}
+
+func (f *FTX) wsResubToOB(p currency.Pair) {
+	channelToResubscribe := wshandler.WebsocketChannelSubscription{
+		Channel:  wsOrderbook,
+		Currency: p,
+	}
+	f.Websocket.ResubscribeToChannel(channelToResubscribe)
+}
+
+// WsProcessPartialOB creates an OB from websocket data
+func (f *FTX) WsProcessPartialOB(data *WsOrderbookData, p currency.Pair, a asset.Item) error {
+	signedChecksum := f.CalcPartialOBChecksum(data)
+	if signedChecksum != data.Checksum {
+		return fmt.Errorf("%s channel: %s. Orderbook partial for %v checksum invalid",
+			f.Name,
+			a,
+			p)
+	}
+	var bids, asks []orderbook.Item
+	for x := range data.Bids {
+		bids = append(bids, orderbook.Item{
+			Price:  data.Bids[x][0],
+			Amount: data.Bids[x][1],
+		})
+	}
+	for x := range data.Asks {
+		asks = append(asks, orderbook.Item{
+			Price:  data.Asks[x][0],
+			Amount: data.Asks[x][1],
+		})
+	}
+
+	newOrderBook := orderbook.Base{
+		Asks:         asks,
+		Bids:         bids,
+		AssetType:    a,
+		LastUpdated:  timestampFromFloat64(data.Time),
+		Pair:         p,
+		ExchangeName: f.Name,
+	}
+
+	if err := f.Websocket.Orderbook.LoadSnapshot(&newOrderBook); err != nil {
+		return err
+	}
+
+	f.Websocket.DataHandler <- wshandler.WebsocketOrderbookUpdate{
+		Exchange: f.Name,
+		Asset:    a,
+		Pair:     p,
+	}
+	return nil
+}
+
+// CalcPartialOBChecksum calculates checksum of partial OB data received from WS
+func (f *FTX) CalcPartialOBChecksum(data *WsOrderbookData) int {
```

Comment:
[ERROR]| 01/06/2020 15:38:19 |routines.go exchange FTX websocket error - FTX channel: futures. Orderbook partial for ALT-PERP checksum invalid
[ERROR]| 01/06/2020 15:38:20 |routines.go exchange FTX websocket error - ob.Base could not be found for Exchange FTX CurrencyPair: ALT-PERP AssetType: futures